File: C++/ps_baekjoon/10423/main.cpp
#include <bits/stdc++.h>
using namespace std;
using vertex = tuple<int, int, int>;

constexpr int mxN = 1e3;
int n, m, k, res, uf[mxN + 1];
priority_queue<vertex, vector<vertex>, greater<vertex>> pq;

int find(int num)
{
    if (num == uf[num])
        return num;
    return uf[num] = find(uf[num]);
}

bool uni(int a, int b)
{
    a = find(a);
    b = find(b);
    if (a != b)
    {
        uf[b] = a;
        return true;
    }
    return false;
}

int main()
{
    cin.tie(0)->sync_with_stdio(0);
    cin >> n >> m >> k;
    for (int i = 1; i <= n; ++i)
        uf[i] = i;
    for (int i = 0; i < k; ++i)
    {
        int plant_now = 0, plant_pre = 0;
        cin >> plant_now;
        uni(plant_now, plant_pre);
        plant_pre = plant_now;
    }
    for (int i = 0; i < m; ++i)
    {
        int u, v, w;
        cin >> u >> v >> w;
        pq.emplace(w, u, v);
    }
    while (!pq.empty())
    {
        auto [cost, vertex1, vertex2] = pq.top();
        pq.pop();
        if (uni(vertex1, vertex2))
            res += cost;
    }
    cout << res;
    return 0;
}

File: Python/ps_baekjoon/10423/main.py
import sys
from queue import PriorityQueue

readline, write = (sys.stdin.readline, sys.stdout.write)
pq = PriorityQueue()
uf = []
vst = []
res = 0


def find(num) -> int:
    if num == uf[num]:
        return num
    if vst[num]:
        vst[uf[num]] = True
    uf[num] = find(uf[num])
    return uf[num]


def uni(a, b) -> bool:
    a, b = find(a), find(b)
    if a != b:
        if vst[a] and vst[b]:
            return False
        if vst[a] ^ vst[b]:
            vst[a], vst[b] = [True] * 2
        uf[a] = b
        return True
    else:
        return False


n, m, k = map(int, readline().split())
plant = [int(i) for i in readline().split()][:k]
uf = [i for i in range(n + 1)]
vst = [True] + [False for _ in range(n)]
for e in plant:
    vst[e] = True
for _ in range(m):
    u, v, w = map(int, readline().split())
    pq.put((w, u, v))
while not pq.empty() and not all(vst):
    cost, vertex1, vertex2 = pq.get()
    if uni(vertex1, vertex2):
        res += cost
write(f"{res}")